### PR TITLE
feat: buffered writer

### PR DESCRIPTION
**Pull request by Gozala**
Implement #69

### CODE REVIEW

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:94]: can you move these to the top? I was confused when I encountered them in code but their declarations hadn't appeared already

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:105]: This'll be out when you go above 24 roots, and then again at 256, cbor will be adding extra bytes to fit in the array length.

You could fix it by adding 1 byte for `>=24` and another byte for `>=256`. There's another boundary at 65,536 where you'd need to add 2 more bytes, but maybe by that time you should be doing a `throw`.

There's also a difference in encoding for different sizes of CIDs for the same boundaries - a tiny CID <24 is going to be 1 byte more compact, then a huge over 256 is going to add another byte. These are going to be uncommon, but not out of the question. Unfortunately your arguments won't account for that, maybe documentation just needs to make it clear that this estimate is only good for sensible defaults.

It seems that the cost of a bad estimate is the huge content shuffle in `close()` which could be invisibly very expensive. Documenting that would be good too.

So, some solid tests for this function for various CIDs might be good!

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:114]:
```suggestion
const totalByteLength = (cids) => {
```

File: lib/buffer-writer.js
Comment (rvagg) [reply lib/buffer-writer.js:105]: I added an [`EncodedLength()`](https://pkg.go.dev/github.com/ipld/go-ipld-prime@v0.16.0/codec/dagcbor#EncodedLength) calculator to the Go dag-cbor, which lets you figure out exactly how many bytes an object will take up once encoded, we could do the same for JS because it's not _too_ hard, but it may not end up being that much cheaper than just doing a dag-cbor encode of a faked object and checking the byte count.

**Comment by rvagg** [comment]
Yeah, this is neat, I like the API I think. I like that you could use the class directly to write into your own buffer too. It's just the issues around size estimation that are a concern. That offset shuffle is a bit nasty and it would be nice if we didn't have to have it since this code is going to so much effort to get it right. In fact I'd also be fine with making the API strictly require you estimate correctly ahead of time to avoid this invisible nastiness entirely.

The other issue I see with mis-estimation is that I you want to use the class directly on your own buffers, then if you overestimate your header size then you get back a subarray that doesn't start at the begining of your buffer - which may be unexpected. Not a big deal but maybe you want to write a CAR directly up against some other data already in your buffer and to avoid the padding that an overestimation will create you'll have to get your estimates spot-on - and it'd be tricky to tell if you got it wrong. Another case for just removing the possibility of getting it wrong - `throw` if you misestimate.

File: lib/buffer-writer.js
Comment (Gozala) [reply lib/buffer-writer.js]: > This'll be out when you go above 24 roots, and then again at 256, cbor will be adding extra bytes to fit in the array length.
>
> You could fix it by adding 1 byte for >=24 and another byte for >=256. There's another boundary at 65,536 where you'd need to add 2 more bytes, but maybe by that time you should be doing a throw.
>

I end up publishing before I did a last push so I'm not sure if that is prior to a fix I have added to account for the `varintSize`.  If so I assume that last change should address this or is that something else ?

> There's also a difference in encoding for different sizes of CIDs for the same boundaries - a tiny CID <24 is going to be 1 byte more compact, then a huge over 256 is going to add another byte. These are going to be uncommon, but not out of the question. Unfortunately your arguments won't account for that, maybe documentation just needs to make it clear that this estimate is only good for sensible defaults.

Yeah I'm not sure if there is a good way to account for all of that. Only thing I have considered is to overestimate a bit, which i think is better than underestimate. Alternatively we could return a range and let user decide which one to go with.

> It seems that the cost of a bad estimate is the huge content shuffle in close() which could be invisibly very expensive. Documenting that would be good too.

I do not think that overhead is huge, as far as I remember browsers optimize case of moving bytes within the same buffer



So, some solid tests for this function for various CIDs might be good!


> I added an [`EncodedLength()`](https://pkg.go.dev/github.com/ipld/go-ipld-prime@v0.16.0/codec/dagcbor#EncodedLength) calculator to the Go dag-cbor, which lets you figure out exactly how many bytes an object will take up once encoded, we could do the same for JS because it's not _too_ hard, but it may not end up being that much cheaper than just doing a dag-cbor encode of a faked object and checking the byte count.

Yeah I was thinking about that as well, but decided it was not worth an effort. However I think it would be great to have one.

**Comment by Gozala** [comment]
> Yeah, this is neat, I like the API I think. I like that you could use the class directly to write into your own buffer too. It's just the issues around size estimation that are a concern. That offset shuffle is a bit nasty and it would be nice if we didn't have to have it since this code is going to so much effort to get it right. In fact I'd also be fine with making the API strictly require you estimate correctly ahead of time to avoid this invisible nastiness entirely.

Yeah moving bytes isn't ideal indeed. Maybe there needs to be a separate API for one use case we have that is:

1. We encode user input into CAR packets of 256MiB or less.
2. We do not know how many roots there will be, as it depends on input

With the current API thinking is you could make a rough guess for amount of roots and start writing, if you have less roots you may be able to make space for more blocks, if you have more roots you may be able to make room for them by reducing number of blocks.

Unfortunately I can not think of any better way to do this, other than writing blocks from the end of the buffer, but that is not great either because order in which we encode is it's natural order. The way I think we may go about it is via single synthetic root linking to desired roots, that way synthetic root could be the last one and we could make pretty accurate estimates.

Anyway if you think it's best to avoid moving bytes in close, I can drop that and leave it up to use space to deal with that code. In that case it would be nice to have a helper function that can at least copy stuff into larger buffer without having to reencode everything on wrong estimate.

**Comment by Gozala** [comment]
> The other issue I see with mis-estimation is that I you want to use the class directly on your own buffers, then if you overestimate your header size then you get back a subarray that doesn't start at the begining of your buffer - which may be unexpected. 

I am not sure it is that bad, although I agree it can be unexpected. Alternatively we could slide bytes blocks towards the head, but I thought I'd leave it up the user to decide. What do you think ?

> Not a big deal but maybe you want to write a CAR directly up against some other data already in your buffer and to avoid the padding that an overestimation will create you'll have to get your estimates spot-on - and it'd be tricky to tell if you got it wrong.

I'm not sure it would be tricky, you just have to compare the `byteOffset`, but I do agree that one may not expect it to be different. For what it's worth that was a reason you pass `ArrayBuffer` and get back `Uint8Array`, hope was that this would make things more clear.

> Another case for just removing the possibility of getting it wrong - throw if you misestimate.

I don't like throwing idea here because you did mutate the buffer but then you threw and left buffer dirty.

I think here are several alternative options to consider:

1. We pad with raw block just to align things.
2. We move blocks instead to align.
3. We move header to the front to align.

I felt like 3rd option was best as it has least overhead and user could still revert back to 1 or 2 if so desired.

File: lib/buffer-writer.js
Comment (rvagg) [reply lib/buffer-writer.js]: > If so I assume that last change should address this or is that something else ?

something else, this is all about cbor encoding compression. Arrays have their length included at the start, if the length is <24 then that number is included in the single byte at the start that says "this is an array". Between 24 and 256 it's in its own byte, so the array takes up 2 bytes to get started, 256 to 65536 takes up another byte, etc. The same rule applies for byte arrays - so CIDs have this property too. A CID that's less than 24 bytes long will have one less byte in the prelude, and a CID greater than 256 bytes long will have an extra one. This is why a size estimator would be handy, because the rules are specific and get a bit weird and surprising in some edges. But the rules are fixes so if you lock them down then you're set.

Unfortunately if you start down this road then you're not far off making a full estimator for arbitrary objects!

So maybe just get this right for obvious cases and not have easy failure modes, and we can get a proper estimator into the dag-cbor codec to use later.

**Comment by rvagg** [comment]
> I don't like throwing idea here because you did mutate the buffer but then you threw and left buffer dirty.

yeah, good point, but that might be an argument for ensuring that you can properly get your header length estimate right, or at least never underestimate. If we can make the API solid in those terms, i.e. if you use it properly then you'll never underestimate, then it becomes a user problem - don't give bad estimation values up-front or you'll end up in a bad place (in which case I still don't mind throwing, you made a boo-boo, consider it programmer error and therefore reasonably fatal).

**Comment by Gozala** [comment]
> yeah, good point, but that might be an argument for ensuring that you can properly get your header length estimate right, or at least never underestimate. If we can make the API solid in those terms, i.e. if you use it properly then you'll never underestimate, then it becomes a user problem - don't give bad estimation values up-front or you'll end up in a bad place (in which case I still don't mind throwing, you made a boo-boo, consider it programmer error and therefore reasonably fatal).

Ok so I have spend bit more time on this and I think I have reasonable solution

1. I have addressed header estimate issues you've caught (thanks for pointing those out)
2. I made `close` throw exception if estimates is incorrect.
3. I have added an option to close `close({ align: true })` which if passed will move bytes when possible and ensure that head starts at the beginning of the buffer and is directly followed by the blocks.

I think that is reasonable compromise, as it fails by default but provides a way to recover when possible.

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:101]:
```suggestion
```

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:166]: We could fix this too now you have the nice `arrayLengthEncodeSize()` which will work for CIDs too; I think if you subtract `2` from `ROOT_EXTRA_SIZE` (the byte array is >24<256 so it requires two bytes cbor prelude `<bytes><length>`, the extra bits in `ROOT_EXTRA_SIZE` are to do with tags, mainly)  and then add it back in by adding `+ arrayLengthEncodeSize(cid.length)` here then you get a more accurate CID length. The problem with doing that is that your API allows for an optional `rootsByteLength`, but that could also probably be fixed by adding those `2` to the `DEFAULT_CID_SIZE`.

If you add `bafkqaaia` and `bafkqbbacaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa` to your tests then you should see the impact of the bad sizes here.

Identity CIDs are the most likely way we're going to end up here--it would be silly to use one as a root but it's allowed and it's not out of the question that someone finds a use-case for it.

**Comment by rvagg** [comment]
👌 nice

re my last comment about CID sizing - it seems like low hanging fruit now you have the logic partially done, but I wouldn't consider it essential; we should just identify this as something to fix later one when we get a proper CBOR sizer and add it in here (I wrote one today for cborg but ran into a nasty bug in the cbor encoder triggered in the test suite that got me side tracked).

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:202]: This argument is still bothering me because it's getting in the way of a more accurate calculation when we can do it.

To illustrate .. and get us closer to actually doing it .. I've added a calculator in https://github.com/rvagg/cborg/pull/48 and then adapted the most efficient form of it against this branch in https://github.com/ipld/js-car/pull/71 (failing because it's not merged and compiled yet but it does pass if wired up).

But you can see from that, that we need individual root lengths, not just an aggregate. An aggregate might be good for this case where we're assuming a standard CID length, but it fails for novelty outside of that.

How would this PR change if we merged #71 (or some variation of it) into here and had an accurate calculator. What API would you present to the user with an accurate calculator but not requiring they provide the CIDs up front?

My preference would be to strip away all the estimation, present the user with an accurate calculator based on lengths and put the onus on them to provide us with either an accurate length, or an estimation that is at least long enough—but it's their responsibility to deal with that.

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:257]: `headerCapacity` here, `estimateHeaderCapacity` below and in test/test-buffer-writer.js seem to be leftovers from a previous iteration that had those names.

File: lib/buffer-writer.js
Comment (Gozala) [reply lib/buffer-writer.js:202]: > But you can see from that, that we need individual root lengths, not just an aggregate. An aggregate might be good for this case where we're assuming a standard CID length, but it fails for novelty outside of that

It is incorrect, however sometimes you don't know how what the roots will look like. So this is meant to just provide estimate. If we'd known the roots ahead we would not need to estimate anything. I also can't imagine case where I'd know CID sizes ahead but not what they are.

Comment (Gozala) [reply lib/buffer-writer.js:202]: > My preference would be to strip away all the estimation, present the user with an accurate calculator based on lengths and put the onus on them to provide us with either an accurate length, or an estimation that is at least long enough—but it's their responsibility to deal with that.

If you want to go with estimator as in #71 that's ok by me as well. In common case you'd just `new Array(count).fill(avgSize)` I suppose.

Comment (Gozala) [reply lib/buffer-writer.js:202]: I could change `estimateHeaderSize` to take `rootLengths` instead even without changes to cborg. I'm also not sure how all those allocations used by estimator will impact performance.

Comment (Gozala) [reply lib/buffer-writer.js:202]: I think at the end of the day I think `estimateHeaderSize` is not a great API if you do know your root sizes, if you're just guessing I'm not sure `calculateHeaderLength` will do any better.

However since one could do:

```js
const estimateHeaderSize = (count, rootsByteLength) =>
  calculateHeaderLength(new Array(count).fill(rootsByteLength / count))
```

But it's not possible to do other way round, I'm convinced that replacing `estimateHeaderSize` with `calculateHeaderLength` is good idea.

Comment (rvagg) [reply lib/buffer-writer.js:202]: > I also can't imagine case where I'd know CID sizes ahead but not what they are

Yeah, but my main problem here is with the second argument and it assumes you do know the sizes. It's foot-gunny and invites bad assumptions. Maybe better to just have the one argument and say that the estimator assumes typical CID sizes, i.e. hide the second argument inside the function, not as an optional argument with a default. _Or_ make the second argument a singular CID length that gets multiplied by the first—by doing that you could account for CIDv0's and CIDs with hashers with multicodecs in the multi-byte range (like Filecoin's blake2b256).

> I'm also not sure how all those allocations used by estimator will impact performance

my concern too, but mocha helps out a bit with measuring that:

```
      ✔ estimateHeaderCapacity(65535) (98ms)
      ✔ calculateHeaderLength(65535) (84ms)
      ✔ estimateHeaderCapacity(65536) (62ms)
      ✔ calculateHeaderLength(65536) (79ms)
```

(very variable timings, they flip around quite a bit, this is a random grab that happens to look good)

(a) the VM is doing a decent job of optimising away some of the allocations (`new` really isn't that scary) and (b) there really aren't that many allocations going on and the call depth involved with the calculator is actually pretty simple, just hidden by abstractions to account for more complex cases, so a lot of that gets optimised away too.

---

What about both? `estimateHeaderSize(count, [cidLength])` for the case where you just want a simple value and know the number of roots and _maybe_ have some novel characteristics about your CIDs, and `calculateHeaderSize(rootLengths)` for the case where you have some certainty.

**Comment by Gozala** [comment]
@rvagg I have merged your #71 and refactored this PR to remove prior calculations. I have also changed  `estimateHeaderSize` to `estimateHeaderLength(rootCount:number, rootByteLength?:number):number` as you've suggested.

I think with that all the issues should be addressed now

File: lib/buffer-writer.js
Comment (rvagg) [inline lib/buffer-writer.js:113]: This appears to be unused, although tested, perhaps remove it to avoid the export noise and API surface area; or if you think this is going to be genuinely useful, add some docs using the pattern I added for the other documented statics (make sure to include a `@name`!).

File: package.json
Comment (rvagg) [inline package.json:58]: you're going to need to add an equivalent block in `typesVersions` so consumers can find the types for it (i.e. avoiding the `lib` indirection)

**Review by rvagg — APPROVED** [review]
Yep, looks good. So I've pushed another commit that compiles the docs into the README - there's a script in package.json that just needed the file added (I add the files manually to control the ordering). Run `npm run docs` to re-generate.

I've named the static members of that file as `CarBufferWriter.*` assuming the pattern is `import * as CarBufferWriter from '@ipld/car/buffer-writer'`, it also makes the TOC cleaner not to have ungrouped static functions.

The logic it uses to pull docs out is that it'll ignore jsdoc blocks where there are no actual docs. When you add some notes above a function it'll appear in the README. Have a look over what I've added and consider whether you want to document more of the exported functions or not or change the docs that are already there.

Giving you :thumbsup: for this so I can switch off today and you can get this merged yourself, a few things though:

* the `typesVersions` addition
* maybe remove `blockEncodeSize`
* any other doc tweaks you think would contribute to making the README more informative
* **please squash away the merge commit** before merging - I don't really mind how many commits this gets merged as, although the commit messages aren't very informative as they are now, I just _really_ don't want a merge commit in master. I'm happy for you to squash away my commits into yours and just make a single `feat:` if you like. I think Squash and Merge will take care of it fine if you want to use that.

**Comment by rvagg** [comment]
oh, the ts compiler doesn't like some of my `@param` additions for docs, those can be removed but it might be nice if you could figure out a compromise to get params documented a little bit nicely

File: lib/buffer-writer.js
Comment (Gozala) [inline lib/buffer-writer.js:224]:
```suggestion
 * @param {object} options
```

File: lib/buffer-writer.js
Comment (Gozala) [inline lib/buffer-writer.js:258]:
```suggestion
 * @param {object} [options]
```

**Pull request merged**
